File: jvm/leetcode_scala/src/main/scala/leetcode/editor/cn/[95]不同的二叉搜索树 II.scala
package leetcode.editor.cn

object UniqueBinarySearchTreesIi {
  import scala.util.control.TailCalls.*
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for a binary tree node. class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) { var value: Int = _value var left: TreeNode = _left var right: TreeNode =
    * _right }
    */
  object Solution {
    def generateTrees(n: Int): List[TreeNode] =
      impl().evalState(Context(1, n))

    private def impl(): State[Context, List[TreeNode]] =
      import StateOps.*
      get[Context].flatMap { ctx =>
        if ctx.low > ctx.high then point(List(null))
        else
          (ctx.low to ctx.high).foldLeft(point(Nil: List[TreeNode])) { (acc, num) =>
            for
              a     <- acc
              left  <- modify[Context](s => s.copy(low = ctx.low, high = num - 1)).flatMap(_ => impl())
              right <- modify[Context](s => s.copy(low = num + 1, high = ctx.high)).flatMap(_ => impl())
            yield (
              for
                l <- left
                r <- right
              yield
                val node = TreeNode(num)
                node.left = l
                node.right = r
                node
            ) ++ a
          }

      }

    private case class Context(low: Int, high: Int)

    import scala.util.control.TailCalls.*

    /** 栈安全的State Monad。可用于leetcode的用fp方式来解决问题。
      */
    case class State[S, A](run: S => TailRec[(S, A)]) {
      def evalState(s: S): A = run(s).result._2

      def execState(s: S): S = run(s).result._1

      def map[B](ab: A => B): State[S, B] =
        flatMap(a => State(s => done((s, ab(a)))))

      def flatMap[B](afb: A => State[S, B]): State[S, B] =
        State(s =>
          tailcall(
            run(s).flatMap { case (s, a) => tailcall(afb(a).run(s)) }
          )
        )
    }

    object StateOps {
      def point[S, A](a: A): State[S, A] = State(s => done((s, a)))

      def get[S]: State[S, S] = State(s => done((s, s)))

      def put[S](s: S): State[S, Unit] = State(_ => done((s, ())))

      def modify[S](ss: S => S): State[S, Unit] =
        for { s <- get[S]; _ <- put(ss(s)) } yield ()
    }

  }
//leetcode submit region end(Prohibit modification and deletion)
}

@main
def runUniqueBinarySearchTreesIi: Unit = {
  import UniqueBinarySearchTreesIi.Solution
  println(Solution.generateTrees(3))
}

File: cpp/leetcode/editor/cn/[95]不同的二叉搜索树 II.cpp
#include <vector>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <utility>
#include <memory>
#include <iterator>
#include <unordered_map>
#include <algorithm>
#include <numeric>

using namespace std;
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        return build(1,n);
    }

    vector<TreeNode*> build(int low,int high) {
        if (low > high) {
            return {nullptr};
        }

        vector<TreeNode*> result;
        for (auto i = low;i<=high;i++) {
            auto left = build(low,i - 1);
            auto right = build(i + 1,high);
            for (auto l : left){
                for (auto r: right){
                    auto t = new TreeNode(i);
                    t->left = l;
                    t->right = r;
                    result.push_back(t);
                }
            }
        }
        return result;
    }
};
//leetcode submit region end(Prohibit modification and deletion)
